### PR TITLE
[X86][MC][CodeGen] Support EGPR for KMOV

### PR DESCRIPTION
**Pull request by KanRobert**
KMOV is essential for copy between k-registers and GPRs.
R16-R31 was added into GPRs in #70958, so we extend KMOV for these new
registers first.

This patch
1.  Promotes KMOV instructions from VEX space to EVEX space
2.  Emits prefix {evex} for the EVEX variants
3.  Prefers EVEX variant than VEX variant in ISEL and optimizations for better RA

EVEX variants will be compressed to VEX variants by existing EVEX2VEX pass if no EGPR is used.

RFC: https://discourse.llvm.org/t/rfc-design-for-apx-feature-egpr-and-ndd-support/73031/4
TAG: llvm-test-suite && CPU2017 can be built with feature egpr successfully.

### CODE REVIEW

File: llvm/lib/Target/X86/X86ExpandPseudo.cpp [hunk 1]
```diff
@@ -466,10 +466,15 @@ bool X86ExpandPseudo::ExpandMI(MachineBasicBlock &MBB,
...
-      .addReg(Reg1, RegState::Define | getDeadRegState(DstIsDead));
+    bool HasEGPR = STI->hasEGPR();
+    auto MIBLo =
+        BuildMI(MBB, MBBI, DL,
+                TII->get(HasEGPR ? X86::KMOVWkm_EVEX : X86::KMOVWkm))
+            .addReg(Reg0, RegState::Define | getDeadRegState(DstIsDead));
+    auto MIBHi =
+        BuildMI(MBB, MBBI, DL,
+                TII->get(HasEGPR ? X86::KMOVWkm_EVEX : X86::KMOVWkm))
+            .addReg(Reg1, RegState::Define | getDeadRegState(DstIsDead));
```
Comment (XinWang10) [inline llvm/lib/Target/X86/X86ExpandPseudo.cpp:477]: In this file, we also have KMOVWmk, do we need to support it too?

File: llvm/lib/Target/X86/X86InstrInfo.td [hunk 2]
```diff
@@ -878,6 +878,8 @@ def relocImm : ComplexPattern<iAny, 1, "selectRelocImm",
 // X86 Instruction Predicate Definitions.
 def TruePredicate : Predicate<"true">;
 
+def HasEGPR : Predicate<"Subtarget->hasEGPR()">;
+def NoEGPR : Predicate<"!Subtarget->hasEGPR()">;
```
Comment (XinWang10) [inline llvm/lib/Target/X86/X86InstrInfo.td:882]: Do we need to adjust the position of ':'.

File: llvm/lib/Target/X86/X86ExpandPseudo.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86ExpandPseudo.cpp:477]: Good catch!

**Review by phoebewang — APPROVED** [review]
LGTM.

**Review by XinWang10 — APPROVED** [review]
LGTM

**Comment by KanRobert** [comment]
@phoebewang @XinWang10 Thanks for the quick review!

**Pull request merged**